### ISSUE TITLE
Longer time intervals

### ISSUE THREAD

**Issue by mehaase** [dbader/schedule#73]
What do you think about adding support for `scheduler.every().month.do(…)` or `scheduler.every().year.do(…)`? I have a job that needs to run on the first of the month. (I'm currently running it every day, and the job bails if it's not the first of the month.)

**Comment by dbader** [comment]
Hey @mehaase, thanks for bringing this up!

What were your experiences with using a recurring job every day to check for the 1st of the month for this use case?

One thing I'd be worried about with adding monthly and yearly intervals is that it becomes easier for folks to use schedule in a way that's error prone. For example:

* Long intervals between job runs make it easier for things to go wrong. For example a re-deploy leads to a process restart that resets the timer.
* For long intervals we'd probably need support for job persistency in order to make sure we don't run jobs more often than necessary or miss job runs (This relates to https://github.com/dbader/schedule/issues/17)

Of course, one could always use something like `every(30).days.do(...)` to schedule longer intervals. But again I'd be worried that we're encouraging a pattern that is error prone.  For mission-critical stuff that absolutely must run once a year it's probably best to just use cron.

So maybe the pattern we want to encourage everyone to use is what you're doing currently, i.e run a "check-in" job every day and see if we need to trigger a particular task. That way things will continue to work across process restarts etc. Thoughts?

**Comment by WoLfulus** [comment]
what about conditional checks? I wanted something like this months ago and I ended up doing it inside the callback just like @mehaase 

    scheduler.every().day.if(lambda d,t: d.day == 1).do(x)

first day of the month executes the job

I haven't tought about process restarts on the same day tho.

**Comment by mehaase** [comment]
> What were your experiences with using a recurring job every day to check for the 1st of the month for this use case?

I haven't deployed it to production yet, so I can't speak to experience. But I will post back if I have any interesting experiences.

I haven't looked under the hood, so I can't comment on the issues you mentioned. In practice, I think the typical use case for monthly jobs is running things during certain business windows, e.g. first day of month, first day of quarter, last day of month, last day of quarter, etc. So the syntax would ideally have an `on(…)` method to specify which days to run.

I don't have a use case for a yearly job, but I suppose it would be running some kind of annual report or computing annual statistics, in which case the `on(…)` method should specify the day of year.

**Comment by ametersky** [comment]
@WoLfulus @dbader 

does and `if()` method exist in order to do this?
`scheduler.every().day.if(lambda d,t: d.day == 1).do(x)`

I'm not seeing it in the source code.

I'm also trying to execute something on the first of the month, every month, and I'd like to find the most elegant way to do this.

**Comment by Casyfill** [comment]
as I've experienced this pain on long-scheduled jobs going wrong (or, rather, not doing anything which was revealed in a month), I have implemented a cron rebooter shell script, that reboots the schedule every day. While it was a bit of a hassle, I think that sufficiently solves any questions on persistence.

I understand that the whole idea of schedule.py was, to large extend, not to use cron, but I feel i'd rather schedule "scheduler" (sorry) once, so that I can simply change python code afterward.
